packages/lists/src/components/StaticList/StaticList.types.ts

```diff
@@ -0,0 +1,8 @@
+interface IStaticListProps<T> {
+  as?: keyof JSX.IntrinsicElements;
+  items?: ReadonlyArray<T>;
+  children?: (item: T, index: number) => React.ReactNode;
+  className?: string;
```

Comment:
@JasonGore, one thing I'd like to explore is providing a type-safe way for devs to provide a `key` that we can use to assign as the `key` attribute per row.

```diff
diff --git a/packages/lists/src/components/StaticList/StaticList.tsx b/packages/lists/src/components/StaticList/StaticList.tsx
index 724d112aa..bc4094a96 100644
--- a/packages/lists/src/components/StaticList/StaticList.tsx
+++ b/packages/lists/src/components/StaticList/StaticList.tsx
@@ -7,9 +7,9 @@ function defaultRowRenderer<T>(item: T, index: number): JSX.Element {
 
 class StaticList<T> extends React.PureComponent<IStaticListProps<T>> {
   public render(): JSX.Element {
-    const { items = [], as: RootTag = 'ul', className, children = defaultRowRenderer } = this.props;
+    const { items = [], as: RootTag = 'ul', className, children = defaultRowRenderer, rowKey } = this.props;
 
-    return <RootTag className={className}>{items.map(children)}</RootTag>;
+    return <RootTag className={className}>{items.map((item: T, index: number, _items: ReadonlyArray<T>) => children(item, index, rowKey))}</RootTag>;
   }
 }
 
diff --git a/packages/lists/src/components/StaticList/StaticList.types.ts b/packages/lists/src/components/StaticList/StaticList.types.ts
index 420038d7b..e8d16bd34 100644
--- a/packages/lists/src/components/StaticList/StaticList.types.ts
+++ b/packages/lists/src/components/StaticList/StaticList.types.ts
@@ -1,8 +1,9 @@
 interface IStaticListProps<T> {
   as?: keyof JSX.IntrinsicElements;
   items?: ReadonlyArray<T>;
-  children?: (item: T, index: number) => React.ReactNode;
+  children?: (item: T, index: number, key?: keyof T) => React.ReactNode;
   className?: string;
+  rowKey?: keyof T;
 }
 
 export { IStaticListProps };
```

Something like the above to avoid the need for `IObjectWithKey`. When I was experimenting, I found the `rowKey` prop not very useful unless we were to offer a level of abstraction around the row renderer. This was because I ended up having to write the child render function anyway...

```diff
diff --git a/packages/lists/src/components/StaticList/StaticList.tsx b/packages/lists/src/components/StaticList/StaticList.tsx
index 724d112aa..bc4094a96 100644
--- a/packages/lists/src/components/StaticList/StaticList.tsx
+++ b/packages/lists/src/components/StaticList/StaticList.tsx
@@ -7,9 +7,9 @@ function defaultRowRenderer<T>(item: T, index: number): JSX.Element {
 
 class StaticList<T> extends React.PureComponent<IStaticListProps<T>> {
   public render(): JSX.Element {
-    const { items = [], as: RootTag = 'ul', className, children = defaultRowRenderer } = this.props;
+    const { items = [], as: RootTag = 'ul', className, children = defaultRowRenderer, rowKey } = this.props;
 
-    return <RootTag className={className}>{items.map(children)}</RootTag>;
+    return <RootTag className={className}>{items.map((item: T, index: number, _items: ReadonlyArray<T>) => children(item, index, rowKey))}</RootTag>;
   }
 }
 
diff --git a/packages/lists/src/components/StaticList/StaticList.types.ts b/packages/lists/src/components/StaticList/StaticList.types.ts
index 420038d7b..e8d16bd34 100644
--- a/packages/lists/src/components/StaticList/StaticList.types.ts
+++ b/packages/lists/src/components/StaticList/StaticList.types.ts
@@ -1,8 +1,9 @@
 interface IStaticListProps<T> {
   as?: keyof JSX.IntrinsicElements;
   items?: ReadonlyArray<T>;
-  children?: (item: T, index: number) => React.ReactNode;
+  children?: (item: T, index: number, key?: keyof T) => React.ReactNode;
   className?: string;
+  rowKey?: keyof T;
 }
 
 export { IStaticListProps };
diff --git a/packages/lists/src/components/StaticList/examples/StaticList.Example.tsx b/packages/lists/src/components/StaticList/examples/StaticList.Example.tsx
index 74cc44c6b..eedd3f7b8 100644
--- a/packages/lists/src/components/StaticList/examples/StaticList.Example.tsx
+++ b/packages/lists/src/components/StaticList/examples/StaticList.Example.tsx
@@ -10,7 +10,8 @@ function generateItems(count: number): number[] {
   return itemsArray;
 }
 
-const ITEMS: ReadonlyArray<number> = generateItems(25);
+type IPerson = { first: string, last: string };
+const ITEMS: ReadonlyArray<IPerson> = [{ first: 'John', last: 'Doe' }, { first: 'Jane', last: 'Doe' }];
 
 const listClassName = mergeStyles({
   listStyleType: 'none',
@@ -35,7 +36,7 @@ const listClassName = mergeStyles({
 const StaticListExample = () => {
   return (
     <StaticList items={ITEMS} className={listClassName}>
-      {(_item: number, index: number) => <li key={index}>{`Item #${index}`}</li>}
+      {(item: IPerson, _index: number) => <li key={item.first + item.last}>{`${item.first} ${item.last}`}</li>}
     </StaticList>
   );
 };
```

Curious if you or anyone else has thoughts. Devs frequently omit the key or rely on `index` which comes back to haunt them later.

Reply:
I'm not too familiar with using `keys` or why you would want type safety, but now they we are minbared on TS3.3 we can always using [JSX generics](https://mariusschulz.com/blog/typescript-2-9-passing-generics-to-jsx-elements).